pyro/infer/abstract_infer.py

```diff
@@ -137,12 +67,43 @@ def _traces(self, *args, **kwargs):
         """
         raise NotImplementedError("inference algorithm must implement _traces")
 
+    def _marginal_single_site(self, site="_RETURN"):
+        sites_key = frozenset([site])
+        if sites_key in self._marginals:
+            return self._marginals[sites_key]
+        marginal = Empirical()
+        for tr, log_weight in zip(self.exec_traces, self.log_weights):
+            marginal.add(tr.nodes[site]["value"], log_weight=log_weight)
+        self._marginals[sites_key] = marginal
+        return marginal
+
+    def marginal(self, sites=None):
+        """
+        Return the marginal distribution for one or more sample sites from the
+        collection of execution traces. If multiple sites are specified, they
+        must have the same tensor shape. The result is cached for any future
+        calls.
+
+        :param sites: sample site whose marginal to return.
+        :return: ``dist.Empirical``
+        """
+        if sites is None:
+            sites = "_RETURN"
+        assert isinstance(sites, (list, str))
+        sites_key = tuple(sites) if isinstance(sites, list) else sites
+        if sites_key in self._marginals:
+            return self._marginals[sites_key]
```

Comment:
What's the use case motivating caching marginals internally rather than just using explicit references?